lib/components/chat_panel.dart

```diff
@@ -264,15 +264,18 @@ class _ChatPanelWidgetState extends State<ChatPanelWidget>
                   vsync: this,
                   controller: _controller,
                   itemBuilder: (index) => StyleModelTheme(
-                    child: ChatHistoryMessage(
-                        message: messages[index], channel: widget.channel),
-                  ),
+                      child: ChatHistoryMessage(
+                          key: ValueKey(messages[index].messageId),
+                          message: messages[index],
+                          channel: widget.channel)),
                   isPinnedBuilder: (index) {
                     final expiration = expirations[index];
-                    if (expiration != null) {
-                      return expiration.isAfter(now);
+                    if (expiration == null) {
+                      return PinState.notPinnable;
                     }
-                    return false;
```

Comment:
The gist of this improvement is that `null` used to represent a message that could not be pinned. Pinned messages are generally more expensive to layout and paint, but we were returning `false` instead of `null` which signaled that every message could have been pinned, thus dramatically increasing how expensive rendering is.